python/lsst/cp/pipe/cpFlatNormTask.py

```diff
@@ -74,93 +165,207 @@ class CpFlatNormalizationTask(pipeBase.PipelineTask,
+        outputScales : `dict` [`dict` [`dict` [`float`]]]
+            Dictionary of scales, indexed by detector (`int`),
+            amplifier (`int`), and exposure (`int`).
+
+        Raises
+        ------
+        KeyError
+            Raised if the input dimensions do not contain detector and
+            exposure, or if the metadata does not contain the expected
+            statistic entry.
         """
-        # Load gains from camera.
-        gains = dict()
-        measured = dict()
-        outputScales = dict()
-        for detector in camera:
-            if detector.getId() not in gains:
-                gains[detector.getId()] = dict()
-            if detector.getId() not in measured:
-                measured[detector.getId()] = dict()
-                outputScales[detector.getId()] = dict()
-
-            for amplifier in detector:
-                gains[detector.getId()][amplifier.getName()] = amplifier.getGain()
-                measured[detector.getId()][amplifier.getName()] = dict()
-                outputScales[detector.getId()][amplifier.getName()] = dict()
-
-        # If the gains are stable, than differences between exposures
-        # are caused by differences in lamp brightness and vignetting
-        # (ignoring sky/screen illumination issues we can't fix).
-        # Find the incident flux per exposure by averaging the
-        # amp-level measurement divided by the gain.
-        visitLevel = dict()
-        count = dict()
-
-        for inMetadata, inDimensions in zip(inputExps, inputDims):
-            detectorId = inDimensions.get('detector', None)
-            visitId = inDimensions.get('visit', None)
-
-            if visitId not in visitLevel:
-                visitLevel[visitId] = 0.0
-            if visitId not in count:
-                count[visitId] = 0
-
-            for amplifier in detector:
-                ampName = amplifier.getName()
-                value = inMetadata.get(f'ISR BKG {ampName} MEDIAN', np.nan)
+        expSet = sorted(set([d['exposure'] for d in inputDims]))
+        detSet = sorted(set([d['detector'] for d in inputDims]))
+
+        if self.config.level == 'DETECTOR':
+            bgMatrix = np.zeros((len(detSet), len(expSet)))
+            bgCounts = np.ones((len(detSet), len(expSet)))
+        elif self.config.level == 'AMP':
+            nAmp = len(camera[0])
+            bgMatrix = np.zeros((len(detSet) * nAmp, len(expSet)))
+            bgCounts = np.ones((len(detSet) * nAmp, len(expSet)))
+
+        for inMetadata, inDimensions in zip(inputMDs, inputDims):
+            try:
+                exposureId = inDimensions['exposure']
+                detectorId = inDimensions['detector']
+            except Exception as e:
+                raise KeyError("Cannot find expected dimensions in %s %s" % (inDimensions, e))
+
+            if self.config.level == 'DETECTOR':
+                detId = detSet.index(detectorId)
```

Comment:
Is this efficient? I think it would be better to get this from a lookup table mapping `detectorId` to `detId`. Ditto for the same pattern used in multiple places below.